rest/views.py

```diff
@@ -0,0 +1,178 @@
+        """Filter to these repositories"""
+
+        repo_ids = [repo.id for repo in get_repos(self.request.user.id)]
+        return Repository.objects.filter(id__in=repo_ids)
+
+    def perform_create(self, serializer):
+        """Create a new repository"""
+        # pylint: disable=protected-access
+        add_repo_permission = '{}.add_{}'.format(
+            Repository._meta.app_label,
+            Repository._meta.model_name
+        )
+        if not self.request.user.has_perm(add_repo_permission):
+            raise PermissionDenied()
+
+        repo = serializer.save()
+        assign_user_to_repo_group(
+            self.request.user,
+            repo,
+            GroupTypes.REPO_ADMINISTRATOR
+        )
+
+
+class RepositoryDetail(RetrieveAPIView):
+    """REST detail view for Repository"""
+    serializer_class = RepositorySerializer
+    lookup_field = 'slug'
+    lookup_url_kwarg = 'repo_slug'
+
+    def get_queryset(self):
+        """Filter to this repository"""
+        repo_ids = [repo.id for repo in get_repos(self.request.user.id)]
+        return Repository.objects.filter(
+            slug=self.kwargs['repo_slug']
+        ).filter(id__in=repo_ids)
+
+
+class VocabularyList(ListCreateAPIView):
+    """REST list view for Vocabulary"""
+    serializer_class = VocabularySerializer
+    lookup_field = 'slug'
+    lookup_url_kwarg = 'vocab_slug'
+
+    def get_queryset(self):
+        """Filter to vocabularies for a repository"""
+        repo_ids = [repo.id for repo in get_repos(self.request.user.id)]
+        repos = Repository.objects.filter(
+            slug=self.kwargs['repo_slug']
+        ).filter(id__in=repo_ids)
+        if not repos.exists():
+            return []
+
+        return repos.first().vocabulary_set.all()
+
+    def get_success_headers(self, data):
+        """Add Location header for model create"""
+        url = reverse('vocabulary-detail', kwargs={
+            'repo_slug': self.kwargs['repo_slug'],
+            'vocab_slug': data['slug'],
+        })
+        return {'Location': url}
+
+
+class VocabularyDetail(RetrieveUpdateAPIView):
+    """REST detail view for Vocabulary"""
+    serializer_class = VocabularySerializer
+    lookup_field = 'slug'
+    lookup_url_kwarg = 'vocab_slug'
+
+    def get_queryset(self):
+        """Filter to a vocabulary within a repository"""
+        repo_ids = [repo.id for repo in get_repos(self.request.user.id)]
+        repos = Repository.objects.filter(
+            slug=self.kwargs['repo_slug']
+        ).filter(id__in=repo_ids)
+        if not repos.exists():
+            return []
+
+        return repos.first().vocabulary_set.filter(
+            slug=self.kwargs['vocab_slug']
+        )
+
+
+class TermList(ListCreateAPIView):
+    """REST list view for Term"""
+    serializer_class = TermSerializer
+    lookup_field = 'slug'
+    lookup_url_kwarg = 'term_slug'
+
+    def get_queryset(self):
+        """Filter to terms within a vocabulary and repository"""
```

Comment:
same problem of user without permissions here: gets 200 with no results instead of 403